sdks/java/core/src/main/java/org/apache/beam/sdk/coders/MapCoder.java

```diff
@@ -99,10 +100,17 @@ public void encode(
       throw new CoderException("cannot encode a null Map");
     }
     DataOutputStream dataOutStream = new DataOutputStream(outStream);
-    dataOutStream.writeInt(map.size());
+    int size = map.size();
+    dataOutStream.writeInt(size);
+    int i = 0;
     for (Entry<K, V> entry : map.entrySet()) {
       keyCoder.encode(entry.getKey(), outStream, context.nested());
-      valueCoder.encode(entry.getValue(), outStream, context.nested());
+      if (i < size - 1) {
+        valueCoder.encode(entry.getValue(), outStream, context.nested());
+      } else {
+        valueCoder.encode(entry.getValue(), outStream, context);
+      }
+      ++i;
     }
     dataOutStream.flush();
```

Comment:
Flushing is not needed or expected since DataOutputStream is not a buffering output stream.